src/mechanics/mesh/MeshGmsh.cpp

```diff
@@ -0,0 +1,378 @@
+    {
+        throw NuTo::Exception(__PRETTY_FUNCTION__, "Gmsh version 2.0 or higher requiered. - File version is " +
+                                                           std::to_string(header.version));
+    }
+    getline(file, line); // $EndMeshFormat
+    return header;
+}
+
+
+void ReadNodesASCII(std::istream& file, GmshFileContent& fileContent)
+{
+    std::string line;
+    std::getline(file, line);
+    int numNodes = std::stoi(line);
+
+    fileContent.nodes.resize(numNodes);
+    bool is2d = false;
+    bool is3d = false;
+    for (GmshNode& node : fileContent.nodes)
+    {
+        file >> node.id;
+        file >> node.coordinates[0];
+        file >> node.coordinates[1];
+        file >> node.coordinates[2];
+        fileContent.minNodeId = std::min(fileContent.minNodeId, node.id);
+        fileContent.maxNodeId = std::max(fileContent.maxNodeId, node.id);
+        is2d = (is2d || node.coordinates[1] != 0);
+        is3d = (is3d || node.coordinates[2] != 0);
+        std::getline(file, line);
+    }
+    if (is2d)
+        fileContent.dimension = 2;
+    if (is3d)
+        fileContent.dimension = 3;
+
+    std::getline(file, line);
+    std::transform(line.begin(), line.end(), line.begin(), ::toupper);
+    if (line.compare("$ENDNODES") != 0)
+        throw NuTo::Exception(__PRETTY_FUNCTION__, "$EndNodes not found!");
+}
+
+void ReadElementsASCII(std::istream& file, GmshFileContent& fileContent)
+{
+    // first element does not exist. Gmsh uses one based indexing
+    const std::array<unsigned int, 32> elementNumNodesLookUp{0,  2,  3,  4,  4, 8, 6,  5,  3,  6, 9,
+                                                             10, 27, 18, 14, 1, 8, 20, 15, 13, 9, 10,
+                                                             12, 15, 15, 21, 4, 5, 6,  20, 26, 56};
+
+    std::string line;
+    std::getline(file, line);
+    int numElements = std::stoi(line);
+
+    fileContent.elements.resize(numElements);
+    for (GmshElement& element : fileContent.elements)
+    {
+        int numTags;
+        file >> element.id;
+        file >> element.type;
+        file >> numTags;
+
+        element.tags.resize(numTags);
+        for (int& tag : element.tags)
+            file >> tag;
+
+        element.nodes.resize(elementNumNodesLookUp[element.type]);
+        for (int& node : element.nodes)
+            file >> node;
+        std::getline(file, line); // endl
+    }
+    std::getline(file, line);
+    std::transform(line.begin(), line.end(), line.begin(), ::toupper);
+    if (line.compare("$ENDELEMENTS") != 0)
+        throw NuTo::Exception(__PRETTY_FUNCTION__, "$EndElements not found!");
+}
+
+
+void ReadPhysicalNamesASCII(std::istream& file, GmshFileContent& fileContent)
+{
+    std::string line;
+    std::getline(file, line);
+    int numNames = std::stoi(line);
+
+    fileContent.physicalNames.resize(numNames);
+    for (GmshPhysicalNames& name : fileContent.physicalNames)
+    {
+        file >> name.dimension;
```

Comment:
what is the `dimension` of a `name`? Its length? And what is its id? Some continuous number in the vector of all possible names? :stuck_out_tongue:

Reply:
change it to whatever you like. The origin of the naming:
http://gmsh.info/doc/texinfo/gmsh.html#MSH-ASCII-file-format